packages/aserto/src/aserto/client/directory/v2/__init__.py

```diff
@@ -15,27 +15,108 @@
-from aserto.client.directory import NotFoundError
+from aserto.client.directory import ConfigError, NotFoundError
 from aserto.client.directory.v2.helpers import ObjectIdentifier, RelationResponse
 
 
+def build_grpc_channel(address: str, default_address: str, ca_cert_path: str) -> Optional[grpc.Channel]:
+    if address == "" and default_address == "":
+        return None
+        
+    return grpc.secure_channel(
+        target=address or default_address, 
+        credentials=directory.channel_credentials(cert=ca_cert_path),
+    )
+
 class Directory:
     def __init__(
         self,
         *,
-        address: str,
         api_key: str = "",
         tenant_id: str = "",
         ca_cert_path: str = "",
+        address: str = "",
+        reader_address: str = "",
+        writer_address: str = "",
+        importer_address: str = "",
+        exporter_address: str = "",
     ) -> None:
-        self._channel = grpc.secure_channel(
-            target=address, credentials=directory.channel_credentials(cert=ca_cert_path)
-        )
+        validation_error = directory.validate_addresses(address=address, reader_address=reader_address, writer_address=writer_address,
+                              importer_address=importer_address, exporter_address=exporter_address, model_address="")
+        if validation_error is not None:
+            raise validation_error
+
+        self.build_channels(address=address, reader_address=reader_address, writer_address=writer_address, 
+                importer_address=importer_address, exporter_address=exporter_address, ca_cert_path=ca_cert_path)
+
         self._metadata = directory.get_metadata(api_key=api_key, tenant_id=tenant_id)
-        self.reader = reader.ReaderStub(self._channel)
-        self.writer = writer.WriterStub(self._channel)
-        self.importer = importer.ImporterStub(self._channel)
-        self.exporter = exporter.ExporterStub(self._channel)
+        self._reader = (
+            reader.ReaderStub(self._reader_channel)
+            if self._reader_channel is not None
+            else None
+        )
+        self._writer = (
+            writer.WriterStub(self._writer_channel)
+            if self._writer_channel is not None
+            else None
+        )
+        self._importer = (
+            importer.ImporterStub(self._importer_channel)
+            if self._importer_channel is not None
+            else None
+        )
+        self._exporter = (
+            exporter.ExporterStub(self._exporter_channel)
+            if self._exporter_channel is not None
+            else None
+        )
+
+    def build_channels(
+            self, 
+            address: str,
+            reader_address: str,
+            writer_address: str,
+            importer_address: str,
+            exporter_address: str,
+            ca_cert_path: str) -> None:
+        
+        if reader_address != "" and writer_address != "" and importer_address != "" and exporter_address != "":
```

Comment:
I don't really understand the logic in this function? Why is it a special case when none of the addresses are empty?
What I think you want to do here is first build a `Dict[address, List[service_name]]` that groups together services that share an address. The result would look something like this:
```py
{
    "localhost:9393": ["reader", "importer"],
    "localhost:9394": ["writer", "exporter"],
    "localhost:9395": ["model"],
}
```

Then you can create a new channel for each key in the map and assign it to all the services that share the address.

Reply:
In fact, it might be convenient to define a helper class (`Channels`) that would take care of creating and closing the all the channels.